### PR TITLE
[ADD] sql-injection: Add new check to avoid sql injection

### PR DESCRIPTION
**Pull request by moylop260**
- Guideline https://github.com/oca/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection

<s>Note: This is a progressive PR from https://github.com/OCA/pylint-odoo/pull/28</s> (Now is merged!)

Example of comment by @dufresnedavid
screen shot 2016-05-18 at 9 07 28 am

### CODE REVIEW

File: pylint_odoo/test_repo/broken_module/models/broken_model.py [hunk 1]
```diff
@@ -74,3 +74,43 @@ def my_method9(self):
...
+                         (tuple(ids),))
+
+    def sql_injection_method(self, ids):
+        # SQL injection, bad way
+        self.env.cr.execute('SELECT DISTINCT child_id '
+                            'FROM account_account_consol_rel '
+                            'WHERE parent_id IN %s'
+                            % (tuple(ids),))
+
+    def sql_injection_method2(self, ids):
```
Comment (pedrobaeza) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: This one is the same as the method 1. Why putting it?

Comment (moylop260) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: You are right!
Thanks for feedback

Comment (moylop260) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: Fixed

**Comment by pedrobaeza** [comment]
:+1:

**Comment by lescobarvx** [comment]
:+1:

**Comment by oscarolar** [comment]
👍

File: pylint_odoo/test_repo/broken_module/models/broken_model.py [hunk 2]
```diff
@@ -78,3 +78,33 @@ def my_method9(self):
...
+        # This cr.execute2 or cr2.execute should not be detected
+        self._cr.execute2(
+            'SELECT name FROM account WHERE id IN %s' % (tuple(ids),))
+        cr2.execute(
+            'SELECT name FROM account WHERE id IN %s' % (tuple(ids),))
+
+    def sql_injection_method5(self, ids):
+        var = 'SELECT name FROM account WHERE id IN %s'
+        values = ([1, 2, 3, ], )
+        self._cr.execute(var % values)  # sql injection too
```
Comment (dufresnedavid) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: @moylop260 how does this work exactly. I guess these methods will be called somehow in unit tests ? For now, I don't get how you can be sure that sql_method will not be detected and sql_injection_method will be.

Comment (moylop260) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: We have a coverage of 100% here

screen shot 2016-05-18 at 8 55 01 am

Maybe you could be interest in review the following material:

[asciicast](https://asciinema.org/a/26890)

[asciicast](https://asciinema.org/a/26889)

Comment (moylop260) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: You could be the output of the tests [here](https://travis-ci.org/OCA/pylint-odoo/jobs/131135682#L594-L597)

```bash
pylint_odoo/test_repo/broken_module/models/broken_model.py:93: [E8103(sql-injection), TestModel.sql_injection_method] Use of "%" operator in execute database method. Better use parameters instead. - More info https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection
pylint_odoo/test_repo/broken_module/models/broken_model.py:95: [E8103(sql-injection), TestModel.sql_injection_method] Use of "%" operator in execute database method. Better use parameters instead. - More info https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection
pylint_odoo/test_repo/broken_module/models/broken_model.py:97: [E8103(sql-injection), TestModel.sql_injection_method] Use of "%" operator in execute database method. Better use parameters instead. - More info https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection
pylint_odoo/test_repo/broken_module/models/broken_model.py:110: [E8103(sql-injection), TestModel.sql_injection_method5] Use of "%" operator in execute database method. Better use parameters instead. - More info https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection
```

We have a expected number of errors by check [here](https://github.com/OCA/pylint-odoo/pull/29/files#diff-6177c602688f2a88312483059f6523d2R39)
```python
    'sql-injection': 4,
```

I hope to be clear

Comment (dufresnedavid) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: @moylop260 ok, I get it, you must check the logs to verify that your test succeded. I'm ok with that.

**Comment by dufresnedavid** [comment]
:+1:

**Comment by moylop260** [comment]
Could merge it?

File: pylint_odoo/test_repo/broken_module/models/broken_model.py (hunk 2 above)
Comment (moylop260) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py]: Thank you

**Comment by guewen** [comment]
What about:

```python
cr.execute('SELECT id FROM %s' % table_name)
```

Which is valid as we are not injecting a value but relation name (injection using params would not work here).

**Comment by moylop260** [comment]
@guewen 
Thanks for feeback

IMHO here we can use the same criteria like as `sql-commit`
- if we need a `%` instead of params, we should add an explicit comment explaining...
- in this comment you can add `# pylint: disable=sql-injection` to bypass this check.

What do you think?

**Comment by moylop260** [comment]
FYI rebased!

**Comment by moylop260** [comment]
@guewen 
If you don't have a new comment... Could we merge it?

**Comment by moylop260** [comment]
@guewen 
ping
Second call...

**Comment by guewen** [comment]
@moylop260 I was busy for a few days. 

Creative people will use `.format()` instead of `%` or do the substitution before the call. That's what I'd do too to avoid to put the `# pylint: disable=sql-injection` comment.

**Comment by moylop260** [comment]
@guewen 
Thanks for reply.

Yes, you are right!
Currently we don't have detected `.format()` and previous variable assign before the call...
You could add that way for now without disable comment.

**Comment by moylop260** [comment]
@guewen 
Could we merge it?

**Comment by moylop260** [comment]
@pedrobaeza @dreispt 
What about merge it?

**Pull request merged**